app/models/member.rb

```diff
@@ -108,16 +109,19 @@ def user_has_namespace_maintainer_access?(user, namespace)
       effective_access_level(namespace, user) == Member::AccessLevel::MAINTAINER
     end
 
-    def access_level_in_namespace_group_links(user, namespace)
+    def access_level_in_namespace_group_links(user, namespace) # rubocop:disable Metrics/AbcSize
       namespace_group_links = NamespaceGroupLink.for_namespace_and_ancestors(namespace)
                                                 .where(group: user.groups.self_and_descendants).not_expired
 
       if namespace_group_links.count.positive?
         maxlevel_namespace_group_link = namespace_group_links.order(:group_access_level).last
-        membership = Member.for_namespace_and_ancestors(maxlevel_namespace_group_link&.group)
+        membership = Member.for_namespace_and_ancestors(maxlevel_namespace_group_link&.group).not_expired
                      &.where(user:)&.order(:access_level)
 
-        return [maxlevel_namespace_group_link.group_access_level, membership.last.access_level].min
+        return [maxlevel_namespace_group_link.group_access_level, membership.last.access_level].min if membership.any?
+
+        return maxlevel_namespace_group_link.group_access_level
```

Comment:
This return statement is not required here. It will end up giving the member this access level even though their membership has expired in which they are a member.

To test this out:
1) Create a new user
2) Login as admin@email.com
3) Create 2 groups, `Group A`, `Group B`
4) Add the new user to `Group B` with `MAINTAINER` access
5) Invite `Group B` to `Group A` with a group access level of `ANALYST` and do not set an expiration on the group link
6) Logout and login as the user created above
7) Verify you can see Group A and Group B, and that you can access them
8) In rails console expire the membership for the user in `Group B` with `m = Member.last` `m.expires_at = Date.today - 1` `m.save!`
9) You shouldn't see any groups listed on the groups dashboard, however if you visit `http://localhost:3000/group-a` you will still be able to access the `Group A` even though the membership for the user has expired in `Group B` which was linked to `Group A`

Remove this return fixes the issue and the user can no longer access `http://localhost:3000/group-a` as is expected

Please also add a test for this case to ensure that the user cannot access the group through a namespace group link when their membership has expired in their own group which was linked to this group

Reply:
Thank you, I was able to reproduce this bug locally. The expired groups were not listed in the dashboard table, but the user was still able to access them. Please see 3e1d2dc2566a3365559b01de88d62345afee898b for the fix and quick test.

Reply:
Thanks!